_posts/2020-08-24-training-imagenet-with-r/training-imagenet-with-r.Rmd

```diff
@@ -0,0 +1,270 @@
+
+## Preprocessing ImageNet
+
+When dealing with large datasets, even simple tasks like downloading or reading a dataset can be much harder than what you would expect. For instance, since ImageNet is roughly 300GB in size, you will need to make sure we have at least 600GB of free space to leave some room for download and decompression. But no worries, you can always borrow computers with huge disk drives from your favorite cloud provider. While you are at it, you should also request compute instances with multiple GPUs, Solid State Drives (SSDs), and a reasonable amount of CPUs and memory. If you want to use the exact configuration we used, take a look at the [mlverse/imagenet](https://github.com/mlverse/imagenet) repo, which contains a Docker image and configuration commands required to provision reasonable computing resources for this task. In summary, make sure you have access to sufficient compute resources.
+
+Now that we have resources capable of working with ImageNet, we need to find a place to download ImageNet from. Long story short, the easiest way is to use a variation of ImageNet used in the [ImageNet Large Scale Visual Recognition Challenge (ILSVRC)](http://www.image-net.org/challenges/LSVRC/), which contains a subset of about 250GB of data and can be easily downloaded from many [Kaggle](https://kaggle.com) competitions, like the [ImageNet Object Localization Challenge](https://www.kaggle.com/c/imagenet-object-localization-challenge).
+
+If you've read some of our previous posts, you might be thinking of using -- the [pins](https://pins.rstudio.com) package. The pins package enables you to cache, discover and share resources from many services, including Kaggle. You can learn more about data retrieval from Kaggle in the [Using Kaggle Boards](http://pins.rstudio.com/articles/boards-kaggle.html) article; in the meantime, let's assume you are already familiar with this package.
+
+All we need to do now is register the Kaggle board, retrieve ImageNet as a pin, and uncompress this file. Warning, the following code requires you to stare at a progress bar for, potentially, over an hour.
+
+```{r}
+library(pins)
+board_register("kaggle", token = "kaggle.json")
+
+pin_get("c/imagenet-object-localization-challenge", board = "kaggle")[1] %>%
+  untar(exdir = "/localssd/imagenet/")
+```
+
+If we are going to be training this model over-and-over using multiple GPUs and even multiple compute instances, we want to make sure we don't waste too much time downloading ImageNet every single time.
+
+The first improvement to consider is getting a faster hard drive. In our case, we locally-mounted an array of SSDs into the `/localssd` path. We then used `/localssd` to extract ImageNet and configured R's temp paths and pins cache to use the SSDs as well. Consult your cloud provider documentation to configure SSDs or take a look at [mlverse/imagenet](https://github.com/mlverse/imagenet).
```

Comment:
should this be `path` (singular)?

Reply:
If "tak[ing] a look at mlverse/imagenet" is an option that does _not_ involve consulting the cloud provider's documentation, there should be a comma before the _or_. Right now it reads as "consult your cloud provider…to take a look at mlverse/imagenet" (since there's no serial comma)

Reply:
Gotcha, fixing.